crates/neon-macros/src/export/function/mod.rs

```diff
@@ -0,0 +1,187 @@
+                let res = #name(#context_arg #(#args),*);
+                #json_return
+                res
+            })
+            .promise(|mut cx, res| neon::types::extract::TryIntoJs::try_into_js(res, &mut cx));
+
+            Ok(neon::handle::Handle::upcast(&promise))
+        ),
+    };
+
+    // Generate the wrapper function
+    let wrapper_fn = quote::quote!(
+        #[doc(hidden)]
+        fn #wrapper_name(mut cx: neon::context::FunctionContext) -> neon::result::JsResult<neon::types::JsValue> {
+            #call_body
+        }
+    );
+
+    // Generate the function that is registered to create the function on addon initialization.
+    // Braces are included to prevent names from polluting user code.
+    let create_fn = quote::quote!({
+        #[doc(hidden)]
+        #[neon::macro_internal::linkme::distributed_slice(neon::macro_internal::EXPORTS)]
+        #[linkme(crate = neon::macro_internal::linkme)]
+        fn #create_name<'cx>(
+            cx: &mut neon::context::ModuleContext<'cx>,
+        ) -> neon::result::NeonResult<(&'static str, neon::handle::Handle<'cx, neon::types::JsValue>)> {
+            static NAME: &str = #export_name;
+
+            #wrapper_fn
+
+            neon::types::JsFunction::with_name(cx, NAME, #wrapper_name).map(|v| (
+                NAME,
+                neon::handle::Handle::upcast(&v),
+            ))
+        }
+    });
+
+    // Output the original function with the generated `create_fn` inside of it
+    quote::quote!(
+        #(#attrs) *
+        #vis #sig {
+            #create_fn
+            #block
+        }
+    )
+    .into()
+}
+
+// Get the ident for the first argument
+fn first_arg_ident(sig: &syn::Signature) -> Option<&syn::Ident> {
+    let arg = sig.inputs.first()?;
+    let ty = match arg {
+        syn::FnArg::Receiver(v) => &*v.ty,
+        syn::FnArg::Typed(v) => &*v.ty,
+    };
+
+    let ty = match ty {
+        syn::Type::Reference(ty) => &*ty.elem,
+        _ => return None,
+    };
+
+    let path = match ty {
+        syn::Type::Path(path) => path,
+        _ => return None,
+    };
+
+    let path = match path.path.segments.last() {
+        Some(path) => path,
+        None => return None,
+    };
+
+    Some(&path.ident)
+}
+
+// Determine if the function has a context argument and if it is allowed
+fn has_context_arg(meta: &meta::Meta, sig: &syn::Signature) -> syn::Result<bool> {
+    // Return early if no arguments
+    let first = match first_arg_ident(sig) {
+        Some(first) => first,
+        None => return Ok(false),
+    };
+
+    // First argument isn't context
+    if first != "FunctionContext" {
+        return Ok(false);
+    }
+
+    // Context is only allowed for normal functions
+    match meta.kind {
+        Kind::Normal => {}
+        Kind::Task => return Err(syn::Error::new(first.span(), TASK_CX_ERROR)),
+    }
+
+    Ok(true)
+}
+
+// Determine if a return type is a `Result`
+fn is_result_output(ret: &syn::ReturnType) -> bool {
```

Comment:
Also here, the criteria should be spec'ed in the `#[export]` API docs.

Reply:
I documented this, but also added a `result` attribute to mirror the `context` attribute. This allows a user to use type aliases for `Result`.